Local "Sending" member invite & promotion status logic

mpretty-cyro opened the pull request.

mpretty-cyro committed:
Initial version of the `local_pending_send` flag

jagerman commented on include/session/config/groups/members.hpp:
```diff
@@ -287,11 +314,14 @@ struct member {
 class Members : public ConfigBase {
 
+  private:
+    std::optional<std::unordered_set<std::string>> pending_send_ids;
```
```suggestion
    std::optional<std::unordered_set<std::string>> pending_send_ids;
```
I don't think having this be `optional` is useful; what optional gives you is an extra "no data" state, which is useful if you want to distinguish between "this is not set at all" and "this is empty".  But here it seems like empty is perfectly fine, and would simply things.

mpretty-cyro replied:
I wasn't sure if an empty `std::unordered_set<std::string>` would take up memory (though now that I write that, it should be negligible even if it does so is probably just an unnecessary optimisation)

jagerman committed:
extra_data: use dict producer/consumer

Returning a bt_dict is clunky and slow, and increasingly we're moving
away from bt_dict everywhere else in the codebase in favour of
producer/consumer, so do that for extra_data too.

jagerman commented on include/session/config/groups/members.hpp:
```diff
@@ -489,6 +520,33 @@ class Members : public ConfigBase {
     };
+
+    protected:
+        oxenc::bt_dict extra_data() const override {
+            oxenc::bt_dict result;
+            if (pending_send_ids && !pending_send_ids->empty()) {
+                oxenc::bt_list ids = {};
+                for (const auto& sid : *pending_send_ids)
+                    ids.emplace_back(sid);
+
+                result["pending_send_ids"] = ids;
+            }
+            
+            return result;
+        }
```
With the producer change that I just pushed, and the change away from optional, this whole thing ought to be able to become just:
```suggestion
        void extra_data(oxenc::bt_dict_producer&& extra) const override {
            extra.append("pending_send_ids", pending_send_ids);
        }
```

(You very likely need to update oxen-encoding to the latest `dev` branch for that to work, and for that you might need to cherry-pick oxenc compat fix da5a2f3 from my multipart config PR).

jagerman commented on include/session/config/groups/members.hpp:
```diff
@@ -489,6 +520,33 @@ class Members : public ConfigBase {
+            return result;
+        }
+        
+        void load_extra_data([[maybe_unused]] oxenc::bt_dict extra) override {
+            if (auto it = extra.find("pending_send_ids"); it != extra.end()) {
+                if (std::holds_alternative<oxenc::bt_list>(it->second)) {
+                    auto invitations = std::get<oxenc::bt_list>(it->second);
+                    pending_send_ids = {};
+
+                    for (const auto& sid : invitations)
+                        if (std::holds_alternative<std::string>(sid))
+                            pending_send_ids->emplace(std::get<std::string>(sid));
+                }
+            }
+        }
```
And this one can become:
```suggestion
        void load_extra_data(oxenc::bt_dict_consumer&& extra) override {
            if (extra.skip_until("pending_send_ids"))
                extra.consume(pending_send_ids);
        }
```

jagerman commented on src/config/groups/members.cpp:
```diff
@@ -20,16 +20,21 @@ std::optional<member> Members::get(std::string_view pubkey_hex) const {
 
-    auto result = std::make_optional<member>(std::string{pubkey_hex});
-    result->load(*info_dict);
+    auto sid = std::string{pubkey_hex};
+    auto result = std::make_optional<member>(sid);
+    auto is_pending_send = (pending_send_ids && pending_send_ids->find(sid) != pending_send_ids->end());
+    result->load(*info_dict, {{"local_pending_send", (is_pending_send ? 1 : 0)}});
```
I'm thinking that it might be easier to just expose the `pending_send_ids` on the members object (i.e. make it public) so that calling can just query it directly.  Or, alternatively, just provide some simple wrappers to query and set it:
```C++
bool Members::has_pending_send(std::string sessionid) { return pending_send_ids.count(sessionid); }
void Members::set_pending_send(std::string sessionid, bool pending) {
    if (pending)
        pending_send_ids.insert(sessionid);
    else
        pending_send_ids.erase(sessionid);
}
```

Otherwise it feels a bit like we're shoehorning it through the `member` object though it feels a bit out of place there (and would seem to make things more complicated).

(We might have to revisit that if we end up with lots of things like this, but so far it feels like a one-off).

Bilb replied:
I've made those fixes and pushed to Morgan's branch

mpretty-cyro committed:
Addressed code feedback

• Moved the member status function to the members config (allows us to avoid odd code gymnastics)
• Removed the changes to the member type
• Fixed a couple of CI errors

mpretty-cyro committed:
Fixed a couple of build issues

mpretty-cyro committed:
Further tweaks for CI errors

mpretty-cyro committed:
Added the C header changes I had missed

mpretty-cyro committed:
Yet another change I missed bringing across

mpretty-cyro committed:
Function change for CI error

mpretty-cyro committed:
Fixed some broken tests, fixed a bug with erasing a member

jagerman committed:
Add two-step ConfigBase init via default-ctor+init()

load_extra_data currently doesn't work as intended because it gets
invoked from the ConfigBase base class constructor, but this constructor
runs before the derived class (Members) has been constructed, and so the
virtual method doesn't find a derived instance and so always goes to the
do-nothing base class implementation.

This solves it by moving to actual dump+key setup into an `init()`
method that allows subclasses to do a two-step initialization where they
default-construct the ConfigBase, and then manually call `init()` with
the dump data and keys from within their own constructor, where the
load_extra_data() call will work properly.

Bilb committed:
use single letter for keys dump()

Bilb committed:
add pending_state get/set and fix unit tests

Bilb committed:
set_pending_send marks config as needing dump on change

mpretty-cyro committed:
Merge pull request #3 from Bilb/feature/local_members_sending_state

use single letter for keys dump() & set pending_ send_state

mpretty-cyro commented:
I think this should be ready for another review now (with the latest changes everything seems to be working)

Bilb committed:
chore: fixed some typo

Bilb committed:
chore: run ./utils/format.sh
